x/crosschain/keeper/keeper_cross_chain_tx.go

```diff
@@ -167,17 +170,17 @@ func (k Keeper) CctxAllPending(c context.Context, req *types.QueryAllCctxPending
 	}
 	sends := make([]*types.CrossChainTx, 0)
 
-	// now query the previous nonces up to 100 prior to find any pending cctx that we might have missed
+	// query out tx trackers whose nonces fall in the range of [0, NonceLow) that we might have missed.
 	// need this logic because a confirmation of higher nonce will automatically update the p.NonceLow
 	// therefore might mask some lower nonce cctx that is still pending.
-	startNonce := p.NonceLow - 100
-	if startNonce < 0 {
-		startNonce = 0
+	outTxTrackers, err := k.GetAllTrackerByChainBeforeNonce(ctx, int64(req.ChainId), p.NonceLow, req.Pagination)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to query out tx trackers by chain %d before nonce %d", req.ChainId, p.NonceLow))
 	}
-	for i := startNonce; i < p.NonceLow; i++ {
-		res, found := k.GetNonceToCctx(ctx, tss.TssPubkey, int64(req.ChainId), i)
+	for _, tracker := range outTxTrackers {
```

Comment:
one concern is that outTxTrackers is not after consensus; it's probably not a good idea to rely on outTxTrackers here. 

Also, there might be cases that no outTxTracker exists for a "forgotten" pending cctx.

Reply:
True. Trackesr are not trustworthy info.